rfd/0-job-worker.md

```diff
@@ -0,0 +1,229 @@
+
+##### cgroups
+
+The library implementation will be configured at instantiation with limits that can be parsed into values for `cpu.max`, `memory.max` and `io.max`. These values will originate from runtime flags provided to the server with defaults that do not implement limits. All processes will be started in a new cgroup regardless of the values of these limits. Note that for expediency, it is assumed that cgroups v2 are in use on the system running the service.
+
+##### namespaces
+
+The library implementation will be configured such that it knows how to execute the server binary as a child process, with the proper clone and unshare flags to execute in a new pid, mount and network namespace, and call back into the library to execute the given command for the job. This library instance is obviously separate from the main server process, but due to being reexecuted with the same flags, just a different command (`child` instead of `serve`, see CLI UX below), it will retain all the necessary configuration.
+
+##### Job Execution
+
+The server will initially reexecute itself as follows:
+
+```sh
+/proc/self/exe child [serve flags] -- command [args]...
+```
+
+When calling this, it will set the command's `SysProcAttr` with:
+
+```go
+syscall.SysProcAttr{
+    Cloneflags: syscall.CLONE_NEWPID | // New pid namespace
+        syscall.CLONE_NEWNS | // New mount namespace group
+        syscall.CLONE_NEWNET, // New network namespace
+    Unshareflags: syscall.CLONE_NEWNS, // Isolate process mounts from host
+}
+```
+
+In order to provide proper accounting of the job the server will internally track of the execution with:
+
+- buffers for `stdout` and `stderr`
+- process status (e.g. running, complete, error)
+- exit code
+
+Once reexecuted, the `child` command has several jobs to do. It has to remount the `/proc` filesystem, create and configure the cgroup, set the cgroup for the pid and exec the job.
+
+##### Job Status
+
+Job status is extremely simple and only returns a status code and the exit_code. The status code is one of:
+
+- running: the job has not yet completed
+- completed: the job completed successfully
+- error: the job completed with a non-zero exit code
+
+##### Job Output Storage
+
+Output for each job will be maintained, in memory, as long as the worker implementation exists. This poses a potential memory leak, but for the purposes of this challenge, is an acceptable solution. Later designs would likely need to write job output to disk or possibly even remote storage.
+
+##### Job Output Streaming
+
+The worker library streams job output through an `io.ReadCloser`. The server will indicate the end of the stream, if the job has completed, with the `io.EOF` error. The client may, at any time, close the stream to indicate that it is disconnecting. The client must always close the stream when it no longer needs it to prevent memory leaks.
```

Comment:
Can you please indicate more details on how the library will inform the reader that the stream ended and how it will also inform that new data was written to stderr or stdout?

Reply:
I added more detail about how the library will manage the job's output buffer and client streaming internally. I'm still not sure that I've answered what you're looking for though. Take a look and lmk.